presto-main/src/main/java/com/facebook/presto/operator/aggregation/AggregationCompiler.java

```diff
@@ -48,98 +57,69 @@
     {
         requireNonNull(returnType, "returnType is null");
         requireNonNull(argumentTypes, "argumentTypes is null");
-        for (InternalAggregationFunction aggregation : generateAggregationFunctions(clazz)) {
-            if (aggregation.getFinalType().equals(returnType) && aggregation.getParameterTypes().equals(argumentTypes)) {
+        for (BindableAggregationFunction aggregation : generateBindableAggregationFunctions(clazz)) {
+            if (aggregation.getSignature().getReturnType().equals(returnType) &&
+                    aggregation.getSignature().getArgumentTypes().equals(argumentTypes)) {
                 return aggregation;
             }
         }
         throw new IllegalArgumentException(String.format("No method with return type %s and arguments %s", returnType, argumentTypes));
     }
 
-    public List<InternalAggregationFunction> generateAggregationFunctions(Class<?> clazz)
+    public static List<BindableAggregationFunction> generateBindableAggregationFunctions(Class<?> aggregationDefinition)
     {
-        AggregationFunction aggregationAnnotation = clazz.getAnnotation(AggregationFunction.class);
+        AggregationFunction aggregationAnnotation = aggregationDefinition.getAnnotation(AggregationFunction.class);
         requireNonNull(aggregationAnnotation, "aggregationAnnotation is null");
 
-        DynamicClassLoader classLoader = new DynamicClassLoader(clazz.getClassLoader());
+        DynamicClassLoader classLoader = new DynamicClassLoader(aggregationDefinition.getClassLoader());
+
+        ImmutableList.Builder<BindableAggregationFunction> builder = ImmutableList.builder();
 
-        ImmutableList.Builder<InternalAggregationFunction> builder = ImmutableList.builder();
-        for (Class<?> stateClass : getStateClasses(clazz)) {
+        for (Class<?> stateClass : getStateClasses(aggregationDefinition)) {
             AccumulatorStateSerializer<?> stateSerializer = new StateCompiler().generateStateSerializer(stateClass, classLoader);
-            Type intermediateType = stateSerializer.getSerializedType();
-            Method intermediateInputFunction = getIntermediateInputFunction(clazz, stateClass);
-            Method combineFunction = getCombineFunction(clazz, stateClass);
-            AccumulatorStateFactory<?> stateFactory = new StateCompiler().generateStateFactory(stateClass, classLoader);
 
-            for (Method outputFunction : getOutputFunctions(clazz, stateClass)) {
-                for (Method inputFunction : getInputFunctions(clazz, stateClass)) {
+            for (Method outputFunction : getOutputFunctions(aggregationDefinition, stateClass)) {
+                for (Method inputFunction : getInputFunctions(aggregationDefinition, stateClass)) {
                     for (String name : getNames(outputFunction, aggregationAnnotation)) {
```

Comment:
Not that I introduced any new ;-)
btw. I think they are not avoidable as we really need to generate aggregations for:
 - each class
 - each output function
 - each input function
 - each alias